src/python/pants/backend/python/typecheck/mypy/rules.py

```diff
@@ -54,19 +60,65 @@ class MyPyFieldSet(FieldSet):
     sources: PythonSources
 
 
+@dataclass(frozen=True)
+class MyPyPartition:
+    field_set_addresses: FrozenOrderedSet[Address]
+    closure: FrozenOrderedSet[Target]
+    interpreter_constraints: PexInterpreterConstraints
+    python_version_already_configured: bool
+
+
 class MyPyRequest(TypecheckRequest):
     field_set_type = MyPyFieldSet
 
 
-def generate_args(mypy: MyPy, *, file_list_path: str) -> Tuple[str, ...]:
+def generate_argv(
+    mypy: MyPy, *, file_list_path: str, python_version: Optional[str]
+) -> Tuple[str, ...]:
     args = []
     if mypy.config:
         args.append(f"--config-file={mypy.config}")
+    if python_version:
+        args.append(f"--python-version={python_version}")
     args.extend(mypy.args)
     args.append(f"@{file_list_path}")
     return tuple(args)
 
 
+def check_and_warn_if_python_version_configured(
+    *, config: Optional[FileContent], args: Tuple[str, ...]
+) -> bool:
+    ending = (
+        "Normally, Pants would automatically set this for you based on your code's interpreter "
```

Comment:
Before, Pants would no matter what set `python_version`. But Danny and I realized it's important to allow you to override.

In the Pex repo, which is ==2.7 or >=3.5, we run MyPy both with Py2 and Py3 to ensure it all works. We need to allow users to still do that.

Reply:
> Before, Pants would no matter what set python_version. But Danny and I realized it's important to allow you to override.

Why is that? Shouldn't the constraints always be declared on the targets?

This warning doesn't seem actionable... if there is a valid reason to set these options, then a user would need to figure out how to match it with a regex. So:
1) if there is no valid reason, would remove the warning
2) if there is a valid reason, would add an explicit option that folks could set to disable the warning.

Reply:
> We want to allow users to run ./pants typecheck :: then ./pants typecheck :: --mypy-args='--py2' right after.

That doesn't seem like a desirable thing... wouldn't it be better to mark the targets as requiring both 2 and 3, and then having pants run them in parallel?

Reply:
Danny and I talked about that, but it proves really challenging to figure out what the user intended. Let's say your code is `>=3.5`. Surely you don't want MyPy to run once with Py35, once with Py36, once with Py37, etc. If you have 3.5-3.9 all installed, that means 4 runs over the same code.

Now throw in partitioning on top of it. You could end up running MyPy dozens of times on your codebase, when you didn't want that.

We reasoned that the best compromise is for Pants to do the sensible thing of defaulting to running with the minimum Python version, while still providing a safety valve that you can override it. If you weren't using Pants, you would likewise need to override it via sequential runs of MyPy.

Reply:
> Danny and I talked about that, but it proves really challenging to figure out what the user intended. Let's say your code is >=3.5. Surely you don't want MyPy to run once with Py35, once with Py36, once with Py37, etc. If you have 3.5-3.9 all installed, that means 4 runs over the same code.

You would only run dozens of times if there were dozens of distinct ranges in BUILD files... in which case that is probably what you actually intended to happen.

Reply:
> You would only run dozens of times if there were dozens of distinct ranges in BUILD files...

Negative, unless we're talking about something different. What I think you're proposing is that for code that's Py2 _and_ Py3 compatible, Pants automatically runs over that same code twice, once with Py2 and once with Py3.

Now imagine our default constraints of `>=3.6,<4` that we set for everyone. This means that we'd need to run 3.6, 3.7, 3.8, 3.9, 3.10, etc, over the exact same code. That's very bad default behavior.